src/hotspot/share/opto/replacednodes.cpp

```diff
@@ -106,90 +106,142 @@ void ReplacedNodes::apply(Node* n, uint idx) {
-              replace = false;
-              break;
-            }
+        } else if (n->in(0) != nullptr && !n->in(0)->is_top()) {
+          Node* c = n->in(0);
+          if (is_dominator(ctl, c)) {
+            clone_uses_and_replace(C, initial, improved, stack, -1);
           }
         } else {
-          for (DUIterator k = n->outs(); n->has_out(k); k++) {
-            enqueue_use(n, n->out(k), work);
+          uint idx = stack.index();
+          if (idx < n->outcnt()) {
+            stack.set_index(idx + 1);
+            stack.push(n->raw_out(idx), 0);
           }
         }
       }
-      if (replace) {
-        bool is_in_table = C->initial_gvn()->hash_delete(use);
-        int replaced = use->replace_edge(initial, improved);
-        if (is_in_table) {
-          C->initial_gvn()->hash_find_insert(use);
+      if (stack.size() == current_size) {
+        for (;;) {
+          stack.pop();
+          if (stack.is_empty()) {
+            break;
+          }
+          n = stack.node();
+          uint idx = stack.index();
+          if (idx < n->outcnt()) {
+            stack.set_index(idx + 1);
+            stack.push(n->raw_out(idx), 0);
+            break;
+          }
         }
-        C->record_for_igvn(use);
-
-        assert(replaced > 0, "inconsistent");
-        --j;
       }
     }
   }
 }
 
+// Clone all nodes on the stack and replace initial by improved for the use at the bottom of the stack
+void ReplacedNodes::clone_uses_and_replace(Compile* C, Node* initial, Node* improved, const Node_Stack& stack, int i) const {
+  Node* prev = stack.node();
+  for (uint k = stack.size() - 2; k > 0 ; k--) {
+    Node* n = stack.node_at(k);
+    Node* clone = n->clone();
+    bool is_in_table = C->initial_gvn()->hash_delete(prev);
+    if (i == -1) {
+      int replaced = prev->replace_edge(n, clone);
+      assert(replaced > 0, "expected some use");
+    } else {
+      assert(k == (stack.size() - 2) && prev->is_Phi(), "");
+      assert(prev->in(i) == n, "not a use?");
+      prev->set_req(i, clone);
+      i = -1;
+    }
+    C->record_for_igvn(prev);
+    if (is_in_table) {
+      C->initial_gvn()->hash_find_insert(prev);
+    }
+    C->initial_gvn()->set_type_bottom(clone);
+    prev = clone;
+  }
+  bool is_in_table = C->initial_gvn()->hash_delete(prev);
+  if (i == -1) {
+    int replaced = prev->replace_edge(initial, improved);
+    assert(replaced > 0, "expected some use");
+  } else {
+    assert(prev->is_Phi(), "only for Phis");
+    assert(prev->in(i) == initial, "not a use?");
+    prev->set_req(i, improved);
+  }
+  C->record_for_igvn(prev);
+  if (is_in_table) {
+    C->initial_gvn()->hash_find_insert(prev);
+  }
+}
+
+bool ReplacedNodes::is_dominator(const Node* ctl, Node* n) const {
+  assert(n->is_CFG(), "should be CFG now");
+  int depth = 0;
+  while (n != ctl) {
+    n = IfNode::up_one_dom(n);
+    depth++;
+    // limit search depth
+    if (depth >= 100 || n == nullptr) {
+      return false;
```

Comment:
Yes the `100` limit is not new. The inconsistency in this bug is when the input of a `Cast` that's on a if projection is replaced but not the input of the if. The `Cast` node is further away than the `If`. Doing the replacement in the `If` but not the `Cast` wouldn't result in a broken graph. And if replacement happens in the `Cast` then it must happen in the `If` because it's closer to the call that's being replaced.